### DISCUSSION TITLE
Can `zarr` work with PyTorch when `num_workers > 0`?

### DISCUSSION THREAD

**Discussion by adosar** [zarr-developers/zarr-python#2542] [Q&A]
Dear `zarr-python` team,

I am trying to build a PyTorch dataset for point clouds and I was wondering if `zarr` can natively play with PyTorch especially in the case where multiprocessing is used for data loading, i.e. in the case `num_workers > 0`. Since the datasets are generally large, I would like the storing method to have the following properties:
* Lazy loading
* Key-value access
* Fast random access

Currently, I am storing the point clouds as arrays in a `.npz`, i.e. `np.savez(file, key_1=pcd_1, key_2=pcd_2, ...)`. However, I am facing the problem as described [here](https://github.com/numpy/numpy/issues/18124#issuecomment-1603503474).

The proposed solution is to avoid loading the `.npz` file into the main process and **load it individually for each worker**. However, if the number of arrays stored in `.npz` are very large, it takes a significant amount of time to load it with `np.load` (I think the large loading time is due to metadata and not the actual arrays).
 
It is trivial to do the same thing with `zarr` since it has an interface very similar to `np.savez`. I am just not sure if it uses `zipfile` under the hood so I will run again into the same problems.

**Comment by jhamman** [comment]
To your main question, we have found that Zarr works great with pytorch + multiprocessing (https://earthmover.io/blog/cloud-native-dataloader).

> I am just not sure if it uses zipfile under the hood so I will run again into the same problems.

Zarr's storage container is flexible. You can use a zipfile if you want. Or you can use a directory store or cloud object store. You'll have to pick what works for your application.

**Reply by eulerleibniz** [reply]
I have read your blog post @jhamman and for the first time I've learnt about XArray and XBatcher which is great 🙏.
I wonder if I really need those libraries for Zarr to work well with PyTorch? Can i use the standard PyTorch dataset and dataloader logic combined with Zarr and achieve good performance?
I have 1TB data comprising of thousands of arrays with shape (C, T) where C=20 and T ~1e6. I want to concatenate them all together into a single massive Zarr dataset, Then access chunks of around 1MB max (arrays of shape [20, ~1e4]). If i wanted to do this with numpy files that would be 1e6 files which is why i want Zarr or maybe HDF5. All my operations on the data is read-only but its highly parallel, i need multiple workers and large batch sizes which translates to multi-process and multi-threads per process i believe. Is Zarr a good use for this?

**Reply by jhamman** [reply]
You can definitely use Zarr directly without Xarray/Xbatcher. Those tools are useful for managing/slicing into hypercubes but if you data is arranged in an easy to sample way, using Zarr directly should work great.
